ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/ContractBasedBlockHashLookup.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.tuweni.units.bigints.UInt256;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Retrieves block hashes from system contract storage and caches hashes by number, used by
+ * BLOCKHASH operation.
+ */
+public class ContractBasedBlockHashLookup implements BlockHashLookup {
+  private static final Logger LOG = LoggerFactory.getLogger(ContractBasedBlockHashLookup.class);
+  private static final long BLOCKHASH_SERVE_WINDOW = 256L;
+
+  private final ProcessableBlockHeader blockHeader;
+  private final Address contractAddress;
+  private final long historyServeWindow;
+  private final long blockHashServeWindow;
+  private final HashMap<Long, Hash> hashByNumber = new HashMap<>();
+
+  /**
+   * Constructs a ContractBasedBlockHashLookup.
+   *
+   * @param currentBlock current block header being processed.
+   * @param contractAddress the address of the contract storing the history.
+   * @param historyServeWindow the number of blocks for which history should be saved.
+   */
+  public ContractBasedBlockHashLookup(
+      final ProcessableBlockHeader currentBlock,
+      final Address contractAddress,
+      final long historyServeWindow) {
+    this(currentBlock, contractAddress, historyServeWindow, BLOCKHASH_SERVE_WINDOW);
+  }
+
+  /**
+   * Constructs a ContractBasedBlockHashLookup with a specified blockHashServeWindow. This
+   * constructor is only used for testing.
+   *
+   * @param currentBlock current block header being processed.
+   * @param contractAddress the address of the contract storing the history.
+   * @param historyServeWindow the number of blocks for which history should be saved.
+   * @param blockHashServeWindow the number of block for which contract can serve the BLOCKHASH
+   *     opcode.
+   */
+  @VisibleForTesting
+  ContractBasedBlockHashLookup(
+      final ProcessableBlockHeader currentBlock,
+      final Address contractAddress,
+      final long historyServeWindow,
+      final long blockHashServeWindow) {
+    this.blockHeader = currentBlock;
+    this.contractAddress = contractAddress;
+    this.historyServeWindow = historyServeWindow;
+    this.blockHashServeWindow = blockHashServeWindow;
+  }
+
+  @Override
+  public Hash apply(final MessageFrame frame, final Long blockNumber) {
+    final long currentBlockNumber = blockHeader.getNumber();
+    final long minBlockServe = Math.max(0, currentBlockNumber - blockHashServeWindow);
+    if (blockNumber >= currentBlockNumber || blockNumber < minBlockServe) {
+      LOG.debug("failed to read hash from system account for block {}", blockNumber);
+      return ZERO;
+    }
+
+    final Hash cachedHash = hashByNumber.get(blockNumber);
+    if (cachedHash != null) {
+      return cachedHash;
+    }
+
+    final WorldUpdater worldUpdater = frame.getWorldUpdater();
+    Account account = worldUpdater.get(contractAddress);
+    if (account == null) {
+      LOG.error("cannot query system contract {}", contractAddress);
```

Comment:
This a valid scenario, correct? In case a network does not have this contract, should we log a warn/trace instead here?

Reply:
This would be a problem that we are not storing hashes in the system contract storage or we are storing in wrong account. I feel it should be an error as it will cause a consensus issue

Reply:
In normal cases it should not occur as the `PragueBlockHashProcessor` that stores hashes in system contract storage should ship with Pectra and populate the contract storage within a day or so, but I would code defensively against it.

>- serve from memory or as per current designs if maintaining requisite history (full clients for e.g.)

There is always this option though. If I get that sentence right, it mentions that we could fallback to the previous approach of finding hashes from blockchain building. That was something I thought about actually but wasn't sure whether add the previous approach as fallback. But I agree with @shemnon, that this is not totally clear. By fatal, did you mean halting execution if this happens? I would agree with that